arduino-ide-extension/src/browser/contributions/user-fields.ts

```diff
@@ -0,0 +1,149 @@
+          order: '2',
+        })
+      );
+    } else {
+      this.menuActionsDisposables.push(
+        registry.registerMenuNode(
+          ArduinoMenus.SKETCH__MAIN_GROUP,
+          new PlaceholderMenuNode(
+            ArduinoMenus.SKETCH__MAIN_GROUP,
+            // commandId: UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.id,
+            UploadSketch.Commands.UPLOAD_WITH_CONFIGURATION.label,
+            { order: '2' }
+          )
+        )
+      );
+    }
+  }
+
+  private selectedFqbnAddress(): string {
+    const { boardsConfig } = this.boardsServiceProvider;
+    const fqbn = boardsConfig.selectedBoard?.fqbn;
+    if (!fqbn) {
+      return '';
+    }
+    const address =
+      boardsConfig.selectedBoard?.port?.address ||
+      boardsConfig.selectedPort?.address;
+    if (!address) {
+      return '';
+    }
+    return fqbn + '|' + address;
+  }
+
+  private async showUserFieldsDialog(
+    key: string
+  ): Promise<BoardUserField[] | undefined> {
+    const cached = this.cachedUserFields.get(key);
+    // Deep clone the array of board fields to avoid editing the cached ones
+    this.userFieldsDialog.value = (
+      cached ?? (await this.boardsServiceProvider.selectedBoardUserFields())
+    ).map((f) => ({ ...f }));
+    const result = await this.userFieldsDialog.open();
+    if (!result) {
+      return;
+    }
+
+    this.userFieldsSet = true;
+    this.cachedUserFields.set(key, result);
+    return result;
+  }
+
+  async checkUserFieldsDialog(forceOpen: boolean): Promise<boolean> {
+    const key = this.selectedFqbnAddress();
+    if (!key) {
+      return false;
+    }
+    /*
+      If the board requires to be configured with user fields, we want
+      to show the user fields dialog, but only if they weren't already
+      filled in or if they were filled in, but the previous upload failed.
+    */
+    if (
+      !forceOpen &&
+      (!this.boardRequiresUserFields ||
+        (this.cachedUserFields.has(key) && this.userFieldsSet))
+    ) {
+      return true;
+    }
+    const userFieldsFilledIn = Boolean(await this.showUserFieldsDialog(key));
+    return userFieldsFilledIn;
+  }
+
+  checkUserFieldsForUpload(): boolean {
+    // TODO: This does not belong here.
+    // IDE2 should not do any preliminary checks but let the CLI fail and then toast a user consumable error message.
+    if (!this.boardRequiresUserFields || this.getUserFields().length > 0) {
+      this.userFieldsSet = true;
+      return true;
+    }
+    this.messageService.error(
+      nls.localize(
+        'arduino/sketch/userFieldsNotFoundError',
+        "Can't find user fields for connected board"
+      )
+    );
+    this.userFieldsSet = false;
+    return false;
+  }
+
+  getUserFields(): BoardUserField[] {
+    return this.cachedUserFields.get(this.selectedFqbnAddress()) ?? [];
+  }
+
+  isRequired(): boolean {
+    return this.boardRequiresUserFields;
+  }
+
+  notifyFailedWithError(e: Error): void {
+    if (
```

Comment:
`CoreError.UploadFailed.is(e)` would be nicer than parsing the message of the error.